bokehjs/src/coffee/models/plots/plot_canvas.coffee

```diff
@@ -402,7 +402,12 @@ class PlotCanvasView extends Renderer.View
     # reach the frame in time (sometimes) so force an update here for now
     @model.get('frame')._update_mappers()
 
+    # Get context and restore to its initial state. The restoring
+    # Would stricly speaking not be necessary, but in case there is an error
+    # during drawing, restore() won't be called enough. This guards for that.
     ctx = @canvas_view.ctx
+    for i in [0..3]
+      ctx.restore()
```

Comment:
The idea is that this will ensure that the context's state stack is empty. So that we have control (in each iteration) of the exact transformations, even if there where errors in the last draw iteration.

The reason that we now do the `scale(ratio, ratio)` into the draw iteration instead of only when the canvas is resized, it because when blitting the WebGL canvas, we need an unscaled canvas (this was causing the blurring). However, doing the scaling on each iteration introduced this vulnerability of blowing everything up when there is an error during the draw iteration. This fixes that vulnerability.